File: app/src/main/java/com/example/zappalex/popularmovies/MovieDetailActivity.java
package com.example.zappalex.popularmovies;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.support.v4.widget.NestedScrollView;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.zappalex.popularmovies.adapters.ReviewAdapter;
import com.example.zappalex.popularmovies.adapters.VideoAdapter;
import com.example.zappalex.popularmovies.data.MovieContract;
import com.example.zappalex.popularmovies.models.Movie;
import com.example.zappalex.popularmovies.models.Review;
import com.example.zappalex.popularmovies.models.Video;
import com.example.zappalex.popularmovies.utilities.FormatUtils;
import com.example.zappalex.popularmovies.utilities.NetworkUtils;
import com.example.zappalex.popularmovies.utilities.TheMovieDbJsonUtils;
import com.squareup.picasso.Picasso;

import java.lang.reflect.Array;
import java.net.URL;
import java.util.ArrayList;


public class MovieDetailActivity extends AppCompatActivity implements VideoAdapter.VideoAdapterOnClickHandler {

    private static final String EXTRA_SCROLL_POSITION = "scroll_position";
    private static final String VIDEOS_QUERY_BUNDLE_EXTRA = "videos_query";
    private static final String REVIEWS_QUERY_BUNDLE_EXTRA = "reviews_query";
    private static final int VIDEOS_LOADER_ID = 56;
    private static final int REVIEWS_LOADER_ID = 57;

    private NestedScrollView mMovieDetailNestedScrollView;
    private TextView mMovieTitleTextView;
    private ImageView mMoviePosterImg;
    private TextView mMovieDateTextView;
    private TextView mMovieRatingTextView;
    private ImageView mMovieFavoriteImageView;
    private TextView mMovieOverviewTextView;

    private Movie mCurrentMovie;
    private boolean mIsMovieInFavorites = false;

    private ArrayList<Video> mVideoList;
    private VideoAdapter mVideoAdapter;
    private ReviewAdapter mReviewAdapter;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_movie_detail);

        initUiComponents();
        initLayoutAdapters();
        retrieveIntents();

        mIsMovieInFavorites = isMovieInFavorites(queryCurrentMovie(), mCurrentMovie);
        toggleFavoritesHeartImage(mIsMovieInFavorites);
    }

    private void initUiComponents() {
        mMovieDetailNestedScrollView = (NestedScrollView) findViewById(R.id.movie_detail_nested_scroll);
        mMovieTitleTextView = (TextView) findViewById(R.id.tv_title_movie);
        mMoviePosterImg = (ImageView) findViewById(R.id.img_movie_poster);
        mMovieDateTextView = (TextView) findViewById(R.id.tv_movie_date);
        mMovieRatingTextView = (TextView) findViewById(R.id.tv_rating);
        mMovieFavoriteImageView = (ImageView) findViewById(R.id.iv_favorite);
        mMovieOverviewTextView = (TextView) findViewById(R.id.tv_movie_overview);

    }

    private void initLayoutAdapters() {
        RecyclerView videoRecyclerView = (RecyclerView) findViewById(R.id.rv_videos);
        videoRecyclerView.setNestedScrollingEnabled(false);
        LinearLayoutManager videosLinearLayoutManager = new LinearLayoutManager(this);
        videoRecyclerView.setLayoutManager(videosLinearLayoutManager);

        mVideoAdapter = new VideoAdapter(this);
        videoRecyclerView.setAdapter(mVideoAdapter);

        RecyclerView reviewRecyclerView = (RecyclerView) findViewById(R.id.rv_reviews);
        reviewRecyclerView.setNestedScrollingEnabled(false);
        LinearLayoutManager reviewsLinearLayoutManager = new LinearLayoutManager(this);
        reviewRecyclerView.setLayoutManager(reviewsLinearLayoutManager);

        mReviewAdapter = new ReviewAdapter();
        reviewRecyclerView.setAdapter(mReviewAdapter);
    }

    private void retrieveIntents() {
        Intent intentThatStartedActivity = getIntent();
        if (intentThatStartedActivity != null && intentThatStartedActivity.hasExtra(MainActivity.PARCELABLE_MOVIE)) {
            mCurrentMovie = intentThatStartedActivity.getParcelableExtra(MainActivity.PARCELABLE_MOVIE);
            populateViews(mCurrentMovie);

            String videosEndpoint = NetworkUtils.formatVideosEndpointWithId(mCurrentMovie.getId());
            fetchVideosOnlyIfOnline(videosEndpoint);

            String reviewsEndpoint = NetworkUtils.formatReviewsEndpointWithId(mCurrentMovie.getId());
            fetchReviewsOnlyIfOnline(reviewsEndpoint);

        } else {
            Toast.makeText(this, getString(R.string.msg_movie_detail_error), Toast.LENGTH_LONG).show();
        }
    }

    private void populateViews(Movie movie) {
        mMovieTitleTextView.setText(movie.getTitle());
        mMovieDateTextView.setText(FormatUtils.getYearFromDateString(movie.getReleaseDate()));
        mMovieRatingTextView.setText(FormatUtils.getFormattedRating(movie.getUserRating()));
        mMovieOverviewTextView.setText(movie.getOverview());

        String picassoImgUrl = NetworkUtils.buildPicassoUrl(movie.getPosterPath());
        Picasso.with(this).load(picassoImgUrl).into(mMoviePosterImg);
    }

    private void fetchVideosOnlyIfOnline(String videosEndpoint) {
        if (NetworkUtils.isDeviceOnline(this)) {
            Bundle queryBundle = new Bundle();
            queryBundle.putString(VIDEOS_QUERY_BUNDLE_EXTRA, videosEndpoint);

            LoaderManager loaderManager = getSupportLoaderManager();
            loaderManager.initLoader(VIDEOS_LOADER_ID, queryBundle, new VideosCallback());

        } else {
            Toast.makeText(this, getString(R.string.msg_videos_offline), Toast.LENGTH_LONG).show();
        }
    }

    private void fetchReviewsOnlyIfOnline(String reviewsEndpoint) {
        if (NetworkUtils.isDeviceOnline(this)) {
            Bundle queryBundle = new Bundle();
            queryBundle.putString(REVIEWS_QUERY_BUNDLE_EXTRA, reviewsEndpoint);

            LoaderManager loaderManager = getSupportLoaderManager();
            loaderManager.initLoader(REVIEWS_LOADER_ID, queryBundle, new ReviewsCallback());
        } else {
            Toast.makeText(this, getString(R.string.msg_reviews_offline), Toast.LENGTH_LONG).show();
        }
    }

    private class VideosCallback implements LoaderManager.LoaderCallbacks<ArrayList<Video>> {
        @Override
        public Loader<ArrayList<Video>> onCreateLoader(int id, final Bundle args) {
            String videosEndpoint = args.getString(VIDEOS_QUERY_BUNDLE_EXTRA);
            return new VideoAsyncTaskLoader(MovieDetailActivity.this, videosEndpoint);
        }

        @Override
        public void onLoadFinished(Loader<ArrayList<Video>> loader, ArrayList<Video> data) {
            mVideoList = data;
            if (mVideoAdapter != null) {
                mVideoAdapter.setVideoList(data);
            }
        }

        @Override
        public void onLoaderReset(Loader<ArrayList<Video>> loader) {
        }
    }

    private static class VideoAsyncTaskLoader extends AsyncTaskLoader<ArrayList<Video>> {

        private ArrayList<Video> mCachedVideos;
        private String mVideosEndpoint = "";

        public VideoAsyncTaskLoader (Context context, String videosEndpoint){
            super(context);
            mVideosEndpoint = videosEndpoint;
        }

        @Override
        protected void onStartLoading() {
            if (mCachedVideos != null) {
                deliverResult(mCachedVideos);
            } else {
                forceLoad();
            }
        }

        @Override
        public ArrayList<Video> loadInBackground() {
            URL videosRequestUrl = NetworkUtils.buildTmdbUrlWithPathEndpoint(mVideosEndpoint);
            if (videosRequestUrl != null) {
                try {
                    String jsonVideosString = NetworkUtils.getResponseFromHttpUrl(videosRequestUrl);
                    return TheMovieDbJsonUtils.getVideoListFromJsonString(jsonVideosString);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            return null;
        }

        @Override
        public void deliverResult(ArrayList<Video> data) {
            mCachedVideos = data;
            super.deliverResult(data);
        }
    }

    private class ReviewsCallback implements LoaderManager.LoaderCallbacks<ArrayList<Review>> {
        @Override
        public Loader<ArrayList<Review>> onCreateLoader(int id, final Bundle args) {
           String reviewsEndpoint = args.getString(REVIEWS_QUERY_BUNDLE_EXTRA);
            return new ReviewsAsyncTaskLoader(MovieDetailActivity.this, reviewsEndpoint);
        }

        @Override
        public void onLoadFinished(Loader<ArrayList<Review>> loader, ArrayList<Review> data) {
            if (mReviewAdapter != null) {
                mReviewAdapter.setReviewList(data);
            }
        }

        @Override
        public void onLoaderReset(Loader<ArrayList<Review>> loader) {
        }
    }

    private static class ReviewsAsyncTaskLoader extends AsyncTaskLoader<ArrayList<Review>>{

        private ArrayList<Review> mCachedReviews;
        private String mReviewsEndpoint = "";

        public ReviewsAsyncTaskLoader(Context context, String reviewsEndpoint){
            super(context);
            mReviewsEndpoint = reviewsEndpoint;
        }

        @Override
        protected void onStartLoading() {
            super.onStartLoading();

            if(mCachedReviews != null){
                deliverResult(mCachedReviews);
            }else{
                forceLoad();
            }
        }

        @Override
        public ArrayList<Review> loadInBackground() {
            URL reviewsRequestUrl = NetworkUtils.buildTmdbUrlWithPathEndpoint(mReviewsEndpoint);
            if (reviewsRequestUrl != null) {
                try {
                    String jsonReviewsString = NetworkUtils.getResponseFromHttpUrl(reviewsRequestUrl);
                    return TheMovieDbJsonUtils.getReviewListFromJsonString(jsonReviewsString);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
            return null;
        }

        @Override
        public void deliverResult(ArrayList<Review> data) {
            mCachedReviews = data;
            super.deliverResult(data);
        }
    };

    @Override
    public void onCLick(int position) {
        if (mVideoList != null && mVideoList.size() > 0) {
            Uri youtubeUri = NetworkUtils.buildYouTubeUriWithVideoKey(mVideoList.get(position).getKey());
            Intent viewVideoIntent = new Intent(Intent.ACTION_VIEW, youtubeUri);

            if (viewVideoIntent.resolveActivity(getPackageManager()) != null) {
                startActivity(viewVideoIntent);
            }
        }
    }

    private boolean isMovieInFavorites(Cursor favoritesCursor, Movie currentMovie) {

        try {
            while (favoritesCursor.moveToNext()) {
                String cursorMovieId = favoritesCursor.getString(favoritesCursor.getColumnIndex(MovieContract.FavoriteMovieEntry.COLUMN_ID));
                if (currentMovie.getId().equals(cursorMovieId)) {
                    return true;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            favoritesCursor.close();
        }

        return false;
    }

    private Cursor queryCurrentMovie() {
        return getContentResolver().query(MovieContract.FavoriteMovieEntry.CONTENT_URI,
                null,
                null,
                new String[]{mCurrentMovie.getId()},
                null);
    }

    private void toggleFavoritesHeartImage(boolean isFavorite) {
        if (isFavorite) {
            mMovieFavoriteImageView.setImageResource(R.drawable.heart_red);
        } else {
            mMovieFavoriteImageView.setImageResource(R.drawable.heart_grey);
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        int scrollPosition = mMovieDetailNestedScrollView.getScrollY();
        outState.putInt(EXTRA_SCROLL_POSITION, scrollPosition);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        int scrollPosition = savedInstanceState.getInt(EXTRA_SCROLL_POSITION);
        mMovieDetailNestedScrollView.scrollTo(0, scrollPosition);
    }

    public void onFavoriteClick(View view) {
        handleFavoriteClickLogic();
    }

    private void handleFavoriteClickLogic() {
        if (mIsMovieInFavorites) {
            deleteMoveFromFavoritesDb();
            mIsMovieInFavorites = false;
        } else {
            insertMovieIntoFavoritesDb();
            mIsMovieInFavorites = true;
        }
        toggleFavoritesHeartImage(mIsMovieInFavorites);
    }

    private void deleteMoveFromFavoritesDb() {

        String idOfMovieToDelete = mCurrentMovie.getId();
        int favoriteDeleted = 0;

        if (idOfMovieToDelete != null) {
            Uri uriToDelete = MovieContract.FavoriteMovieEntry.CONTENT_URI.buildUpon().appendPath(mCurrentMovie.getId()).build();
            favoriteDeleted = getContentResolver().delete(uriToDelete, null, null);
        }

        if (favoriteDeleted == 0) {
            Toast.makeText(this, getString(R.string.action_movie_deleted_from_favorites_failure), Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, getString(R.string.action_movie_deleted_from_favorites), Toast.LENGTH_SHORT).show();
        }
    }

    private void insertMovieIntoFavoritesDb() {
        ContentValues favoriteMovieContentValues = createFavoriteMovieContentValues();
        Uri insertResultUri = null;

        if (favoriteMovieContentValues != null) {
            insertResultUri = getContentResolver().insert(MovieContract.FavoriteMovieEntry.CONTENT_URI, favoriteMovieContentValues);
        }

        if (insertResultUri != null) {
            displayStatusOfSuccessfulInsert(true);
        } else {
            displayStatusOfSuccessfulInsert(false);
        }
    }

    private ContentValues createFavoriteMovieContentValues() {
        if (mCurrentMovie != null) {
            ContentValues cv = new ContentValues();
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_ID, mCurrentMovie.getId());
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_TITLE, mCurrentMovie.getTitle());
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_POSTER_PATH, mCurrentMovie.getPosterPath());
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_OVERVIEW, mCurrentMovie.getOverview());
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_USER_RATING, mCurrentMovie.getUserRating());
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_RELEASE_DATE, mCurrentMovie.getReleaseDate());
            return cv;
        } else {
            return null;
        }
    }

    private void displayStatusOfSuccessfulInsert(boolean isInsertSuccessful) {
        if (isInsertSuccessful) {
            Toast.makeText(this, getString(R.string.action_movie_added_to_favorites_success), Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(this, getString(R.string.action_movie_added_to_favorites_failure), Toast.LENGTH_LONG).show();
        }
    }

}

File: app/src/main/java/com/example/zappalex/popularmovies/MovieDetailActivity.kt
package com.example.zappalex.popularmovies

import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.support.v4.app.LoaderManager
import android.support.v4.content.AsyncTaskLoader
import android.support.v4.content.Loader
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast

import com.example.zappalex.popularmovies.adapters.ReviewAdapter
import com.example.zappalex.popularmovies.adapters.VideoAdapter
import com.example.zappalex.popularmovies.data.MovieContract
import com.example.zappalex.popularmovies.models.Movie
import com.example.zappalex.popularmovies.models.Review
import com.example.zappalex.popularmovies.models.Video
import com.example.zappalex.popularmovies.utilities.FormatUtils
import com.example.zappalex.popularmovies.utilities.NetworkUtils
import com.example.zappalex.popularmovies.utilities.TheMovieDbJsonUtils
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_movie_detail.*
import kotlinx.android.synthetic.main.video_list_item.*

import java.util.ArrayList


class MovieDetailActivity : AppCompatActivity(), VideoAdapter.VideoAdapterOnClickHandler {

    private var currentMovie: Movie? = null
    private var isMovieInFavorites = false

    private var videoList: ArrayList<Video> = arrayListOf<Video>()
    private val videoAdapter = VideoAdapter(this)
    private val reviewAdapter = ReviewAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_movie_detail)

        initLayoutAdapters()
        retrieveIntents()

        isMovieInFavorites = isMovieInFavorites(queryCurrentMovie(), currentMovie)
        toggleFavoritesHeartImage(isMovieInFavorites)

        favoriteImageView.setOnClickListener { handleFavoriteClickLogic() }
    }

    private fun initLayoutAdapters() {

        val videosLinearLayoutManager = LinearLayoutManager(this)
        videosRecyclerView.layoutManager = videosLinearLayoutManager
        videosRecyclerView.adapter = videoAdapter

        val reviewsLinearLayoutManager = LinearLayoutManager(this)
        reviewsRecyclerView.layoutManager = reviewsLinearLayoutManager
        reviewsRecyclerView.adapter = reviewAdapter

        videosRecyclerView.isNestedScrollingEnabled = false
        reviewsRecyclerView.isNestedScrollingEnabled = false
    }

    private fun retrieveIntents() {
        val intentThatStartedActivity = intent
        if (intentThatStartedActivity.hasExtra(MainActivity.PARCELABLE_MOVIE)) {
            currentMovie = intentThatStartedActivity.getParcelableExtra(MainActivity.PARCELABLE_MOVIE)
            currentMovie?.let { populateViews(it) }

            val videosEndpoint = NetworkUtils.formatVideosEndpointWithId(currentMovie?.id)
            fetchVideosOnlyIfOnline(videosEndpoint)

            val reviewsEndpoint = NetworkUtils.formatReviewsEndpointWithId(currentMovie?.id)
            fetchReviewsOnlyIfOnline(reviewsEndpoint)

        } else {
            Toast.makeText(this, getString(R.string.msg_movie_detail_error), Toast.LENGTH_LONG).show()
        }
    }

    private fun populateViews(movie: Movie) {
        movieTitleTextView.text = movie.title
        movieDateTextView.text = FormatUtils.getYearFromDateString(movie.releaseDate)
        ratingTextView.text = FormatUtils.getFormattedRating(movie.userRating)
        movieOverviewTextView.text = movie.overview

        val picassoImgUrl = NetworkUtils.buildPicassoUrl(movie.posterPath)
        Picasso.with(this).load(picassoImgUrl).into(moviePosterImg)
    }

    private fun fetchVideosOnlyIfOnline(videosEndpoint: String) {
        if (NetworkUtils.isDeviceOnline(this)) {
            val queryBundle = Bundle()
            queryBundle.putString(VIDEOS_QUERY_BUNDLE_EXTRA, videosEndpoint)

            val loaderManager = supportLoaderManager
            loaderManager.initLoader(VIDEOS_LOADER_ID, queryBundle, VideosCallback())

        } else {
            Toast.makeText(this, getString(R.string.msg_videos_offline), Toast.LENGTH_LONG).show()
        }
    }

    private fun fetchReviewsOnlyIfOnline(reviewsEndpoint: String) {
        if (NetworkUtils.isDeviceOnline(this)) {
            val queryBundle = Bundle()
            queryBundle.putString(REVIEWS_QUERY_BUNDLE_EXTRA, reviewsEndpoint)

            val loaderManager = supportLoaderManager
            loaderManager.initLoader(REVIEWS_LOADER_ID, queryBundle, ReviewsCallback())
        } else {
            Toast.makeText(this, getString(R.string.msg_reviews_offline), Toast.LENGTH_LONG).show()
        }
    }

    private inner class VideosCallback : LoaderManager.LoaderCallbacks<ArrayList<Video>> {
        override fun onCreateLoader(id: Int, args: Bundle): Loader<ArrayList<Video>> {
            val videosEndpoint = args.getString(VIDEOS_QUERY_BUNDLE_EXTRA)
            return VideoAsyncTaskLoader(this@MovieDetailActivity, videosEndpoint)
        }

        override fun onLoadFinished(loader: Loader<ArrayList<Video>>, data: ArrayList<Video>) {
            videoList = data
            videoAdapter.setVideoList(data)
        }

        override fun onLoaderReset(loader: Loader<ArrayList<Video>>) {}
    }

    private class VideoAsyncTaskLoader(context: Context, val videosEndpoint: String) : AsyncTaskLoader<ArrayList<Video>>(context) {
        private var cachedVideos= arrayListOf<Video>()

        override fun onStartLoading() {
            if (!cachedVideos.isEmpty()) {
                deliverResult(cachedVideos)
            } else {
                forceLoad()
            }
        }

        override fun loadInBackground(): ArrayList<Video> {
            val videosRequestUrl = NetworkUtils.buildTmdbUrlWithPathEndpoint(videosEndpoint)
            videosRequestUrl?.let {
                try {
                    val jsonVideosString = NetworkUtils.getResponseFromHttpUrl(it)
                    return TheMovieDbJsonUtils.getVideoListFromJsonString(jsonVideosString)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
            return arrayListOf<Video>()
        }

        override fun deliverResult(data: ArrayList<Video>) {
            cachedVideos = data
            super.deliverResult(data)
        }
    }

    private inner class ReviewsCallback : LoaderManager.LoaderCallbacks<ArrayList<Review>> {
        override fun onCreateLoader(id: Int, args: Bundle): Loader<ArrayList<Review>> {
            val reviewsEndpoint = args.getString(REVIEWS_QUERY_BUNDLE_EXTRA)
            return ReviewsAsyncTaskLoader(this@MovieDetailActivity, reviewsEndpoint)
        }

        override fun onLoadFinished(loader: Loader<ArrayList<Review>>, data: ArrayList<Review>) {
            reviewAdapter.setReviewList(data)
        }

        override fun onLoaderReset(loader: Loader<ArrayList<Review>>) {}
    }

    private class ReviewsAsyncTaskLoader(context: Context, val reviewsEndpoint: String) : AsyncTaskLoader<ArrayList<Review>>(context) {
        private var cachedReviews= arrayListOf<Review>()

        override fun onStartLoading() {
            super.onStartLoading()

            if (!cachedReviews.isEmpty()) {
                deliverResult(cachedReviews)
            } else {
                forceLoad()
            }
        }

        override fun loadInBackground(): ArrayList<Review> {
            // TODO: fix returning null here?
            val reviewsRequestUrl = NetworkUtils.buildTmdbUrlWithPathEndpoint(reviewsEndpoint)
                try {
                    val jsonReviewsString = NetworkUtils.getResponseFromHttpUrl(reviewsRequestUrl)
                    return TheMovieDbJsonUtils.getReviewListFromJsonString(jsonReviewsString)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            return arrayListOf<Review>()
        }

        override fun deliverResult(data: ArrayList<Review>) {
            super.deliverResult(data)
            cachedReviews = data
        }
    }

    override fun onCLick(position: Int) {
        if (videoList.size > 0) {
            val youtubeUri = NetworkUtils.buildYouTubeUriWithVideoKey(videoList[position].key)
            val viewVideoIntent = Intent(Intent.ACTION_VIEW, youtubeUri)

            if (viewVideoIntent.resolveActivity(packageManager) != null) {
                startActivity(viewVideoIntent)
            }
        }
    }

    private fun isMovieInFavorites(favoritesCursor: Cursor, currentMovie: Movie?): Boolean {
        try {
            while (favoritesCursor.moveToNext()) {
                val cursorMovieId = favoritesCursor.getString(favoritesCursor.getColumnIndex(MovieContract.FavoriteMovieEntry.COLUMN_ID))
                if (currentMovie?.id == cursorMovieId) {
                    return true
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            favoritesCursor.close()
        }

        return false
    }

    private fun queryCurrentMovie(): Cursor {
        return contentResolver.query(MovieContract.FavoriteMovieEntry.CONTENT_URI, null, null,
                arrayOf(currentMovie?.id), null)
    }

    private fun toggleFavoritesHeartImage(isFavorite: Boolean) {
        if (isFavorite) {
            favoriteImageView.setImageResource(R.drawable.heart_red)
        } else {
            favoriteImageView.setImageResource(R.drawable.heart_grey)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        val scrollPosition = movieDetailScrollView.scrollY
        outState.putInt(EXTRA_SCROLL_POSITION, scrollPosition)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)

        val scrollPosition = savedInstanceState.getInt(EXTRA_SCROLL_POSITION)
        movieDetailScrollView.scrollTo(0, scrollPosition)
    }

    fun onFavoriteClick() {
        handleFavoriteClickLogic()
    }

    private fun handleFavoriteClickLogic() {
        if (isMovieInFavorites) {
            deleteMoveFromFavoritesDb()
            isMovieInFavorites = false
        } else {
            insertMovieIntoFavoritesDb()
            isMovieInFavorites = true
        }
        toggleFavoritesHeartImage(isMovieInFavorites)
    }

    private fun deleteMoveFromFavoritesDb() {
        val idOfMovieToDelete = currentMovie?.id
        var favoriteDeleted = 0

        idOfMovieToDelete.let {
            val uriToDelete = MovieContract.FavoriteMovieEntry.CONTENT_URI.buildUpon().appendPath(it).build()
            favoriteDeleted = contentResolver.delete(uriToDelete, null, null)
        }

        if (favoriteDeleted == 0) {
            Toast.makeText(this, getString(R.string.action_movie_deleted_from_favorites_failure), Toast.LENGTH_SHORT).show()
        } else {
            Toast.makeText(this, getString(R.string.action_movie_deleted_from_favorites), Toast.LENGTH_SHORT).show()
        }
    }

    private fun insertMovieIntoFavoritesDb() {
        val favoriteMovieContentValues = createFavoriteMovieContentValues()
        var insertResultUri: Uri? = contentResolver.insert(MovieContract.FavoriteMovieEntry.CONTENT_URI, favoriteMovieContentValues)

        if (insertResultUri != null) {
            displayStatusOfSuccessfulInsert(true)
        } else {
            displayStatusOfSuccessfulInsert(false)
        }
    }

    private fun createFavoriteMovieContentValues(): ContentValues? {
            val cv = ContentValues()
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_ID, currentMovie?.id)
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_TITLE, currentMovie?.title)
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_POSTER_PATH, currentMovie?.posterPath)
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_OVERVIEW, currentMovie?.overview)
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_USER_RATING, currentMovie?.userRating)
            cv.put(MovieContract.FavoriteMovieEntry.COLUMN_RELEASE_DATE, currentMovie?.releaseDate)
            return cv

    }

    private fun displayStatusOfSuccessfulInsert(isInsertSuccessful: Boolean) {
        if (isInsertSuccessful) {
            Toast.makeText(this, getString(R.string.action_movie_added_to_favorites_success), Toast.LENGTH_LONG).show()
        } else {
            Toast.makeText(this, getString(R.string.action_movie_added_to_favorites_failure), Toast.LENGTH_LONG).show()
        }
    }

    companion object {

        private val EXTRA_SCROLL_POSITION = "scroll_position"
        private val VIDEOS_QUERY_BUNDLE_EXTRA = "videos_query"
        private val REVIEWS_QUERY_BUNDLE_EXTRA = "reviews_query"
        private val VIDEOS_LOADER_ID = 56
        private val REVIEWS_LOADER_ID = 57
    }

}
